### DISCUSSION TITLE
(Apache Configuration) How to link postgrest api with my domain name using https?

### DISCUSSION THREAD

**Discussion by dmiepub** [PostgREST/postgrest#1914] [Q&A] [answered]
How to link postgrest api with my domain name using https? if apache 2.4 web server is running.

Pls suggest.

**Comment by colearendt** [comment, marked as answer]
Hi @dmiepub ! 

There are a few pieces of this process that you will want to implement:

1. Terminating HTTPS with your Apache proxy. This is alluded to [here](https://postgrest.org/en/v8.0/admin.html?highlight=ssl#https) (but talks about nginx). Apache has plenty of its own documentation and examples for this - this is the first tutorial I found in a search 😄 It seems decent enough: https://techexpert.tips/apache/enable-https-apache/

1. Postgrest needs to be configured and running somewhere. Let's say it is on port 3000 on the same server where apache is running.

1. Your apache server needs to pass the traffic to postgrest. If your website / domain is already pointed at your apache proxy and you are serving a website from there, then you will want to choose a path to be "forwarded" to postgrest. I.e. `__api__/`. This way your traffic to postgrest is "namespaced" and does not conflict with other routes (like `/index.html`). This is called a reverse proxy, and has plenty of guidance out on the web as well. Writing these rules and deciding on paths, etc. is all up to you. If there is nothing else being served by your proxy, then you can set up _all_ traffic to forward to postgrest by using the base path of `/`. It's also worth noting that you can have multiple domains pointing at your webserver - this often simplifies routing more than using paths like above. For instance, `api.mydomain.com` will send traffic to postgrest, while `www.mydomain.com` will resolve to the website. All such rules can be written in apache (be warned - this step is probably the most complicated. It takes a bit of testing to ensure you've got a good reverse proxy setup! I like to use `curl -i https://mydomain.com/somepath` to test).

    If you used my example from step 2, then apache will forward to `http://locallhost:3000` where Postgrest is running. If you use a path prefix (i.e. `__api__/`), then you will need to be sure the path is "rewritten" / stripped so that `__api__/sometable` becomes `/sometable`. Postgrest listens at the "root" of the domain.

    Some example tutorials:
        - https://techexpert.tips/apache/apache-proxy-configuration/
        - https://httpd.apache.org/docs/2.4/howto/reverse_proxy.html

1. You need to ensure your domain name resolves to your apache server. This is handled by configuring DNS with your domain registrar and pointing it at an address that will get traffic sending to your server / apache. Again, some testing is in order to ensure that traffic is forwarding correctly! `dig mydomain.com` often gives you some good information there. If you want to test the other steps without configuring this step, you can always set up fake dns resolution for yourself in `/etc/host` so that you can test "privately" without configuring anything public over the internet.

I hope that helps! It's a bit of a complex process, but definitely do-able! And there are lots of good resources for this type of thing out there on the web 😄

**Comment by swuecho** [comment]
if you want to try envoy, this is an example https://farcaller.medium.com/how-to-configure-https-backends-in-envoy-b446727b2eb3.

just route the traffic to postgrest instance is ok.

**Comment by GeoGyro** [comment]
Hi @colearendt !

Thanks for your answer and explanations. 

You mentionned good ressources for configuration on the web, do you know where to find examples of configuration files ? I'm trying to deploy it on my domain, especially point 3, route traffic to `postgREST`, [documentation](https://postgrest.org/en/v8.0/admin.html#hardening-postgrest) is a bit short.
